grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractCentralityQuery.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016-2018 Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.graql.internal.query.analytics;
+
+import ai.grakn.concept.Label;
+import ai.grakn.graql.internal.util.StringConverter;
+import com.google.common.collect.Sets;
+
+import javax.annotation.CheckReturnValue;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.joining;
+
+abstract class AbstractCentralityQuery extends AbstractComputeQuery<Map<Long, Set<String>>> {
+
+    /**
+     * The centrality measures supported.
+     */
+    enum CentralityMeasure {
+
+        DEGREE("degree"),
+        K_CORE("k-core");
+        private final String name;
+
+        CentralityMeasure(String name) {
+            this.name = name;
+        }
+
+        @CheckReturnValue
+        public String getName() {
+            return name;
+        }
+    }
+
+    private boolean ofTypeLabelsSet = false;
+    Set<Label> ofLabels = new HashSet<>();
+
+    void initSubGraph() { //TODO: REMOVE THIS METHOD
+        includeAttribute = true;
+    }
+
+    AbstractCentralityQuery setOfTypeLabels(String... ofTypeLabels) {
+        return setOfTypeLabels(Arrays.stream(ofTypeLabels).map(Label::of).collect(Collectors.toSet()));
+    }
+
+    AbstractCentralityQuery setOfTypeLabels(Collection<Label> ofLabels) {
+        if (!ofLabels.isEmpty()) {
+            ofTypeLabelsSet = true;
+            this.ofLabels = Sets.newHashSet(ofLabels);
+        }
+        return this;
+    }
+
+    abstract CentralityMeasure getMethod();
+
+    @Override
+    String graqlString() {
+        String string = "centrality";
+        if (ofTypeLabelsSet) {
+            string += " of " + ofLabels.stream()
+                    .map(StringConverter::typeLabelToString)
+                    .collect(joining(", "));
+        }
+        string += subtypeString() + " using " + getMethod().getName();
+
+        return string;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
These types of equality and hashcode makes me wonder if this class would be a good candidate for `@AutoValue`. If the sets, labels, etc. . . are never mutated I would consider it

Reply:
I thought the same, but it looks like the fields like `ofLabels` do get modified.

Reply:
The user can modify these values by calling the same method twice, e.g. `compute().count().in("person").in("dog");`

Reply:
Awwwww. . . .